File: DumaMarian/Heap Sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void heapify(vector<int>& v, int n, int i)
{
	int largest = i;
	int left = 2 * i + 1;
	int right = 2 * i + 2;
	if (left < n && v[left] > v[largest])
		largest = left;
	if (right < n && v[right] > v[largest])
		largest = right;
	if (largest != i)
	{
			swap(v[i], v[largest]);
			heapify(v, n, largest);
	}
}

void heapSort(vector<int>& v)
{
	int n = v.size();
	for (int i = n / 2 - 1; i >= 0; i--)
		heapify(v, n, i);
	for (int i = n - 1; i > 0; i--)
	{
		swap(v[0], v[i]);
		heapify(v, i, 0);
	}
}

void afisare(vector<int> v)
{
	for (int i = 0; i < v.size(); i++)
		cout << v[i] << " ";
}
int main()
{
	vector<int> v = { 9, 2, 5, 1, 9, 4, 29, 12, 42, 4, 5 };
	heapSort(v);
	afisare(v);
	return 0;
}

File: heap_sort.py
def heapify(lista, n, i):
    """
    Funcția heapify este folosită pentru a transforma o sublistă într-un heap.
    n - dimensiunea listei
    i - indicele nodului care trebuie "heapificat".
    """
    largest = i  
    left = 2 * i + 1  
    right = 2 * i + 2  

    
    if left < n and lista[left] > lista[largest]:
        largest = left

    
    if right < n and lista[right] > lista[largest]:
        largest = right

    
    if largest != i:
        lista[i], lista[largest] = lista[largest], lista[i]  
        heapify(lista, n, largest)  

def heap_sort(lista):
    """
    Funcția heap_sort sortează o listă folosind algoritmul Heap Sort.
    """
    n = len(lista)

    
    for i in range(n // 2 - 1, -1, -1):
        heapify(lista, n, i)

    
    for i in range(n - 1, 0, -1):
        lista[i], lista[0] = lista[0], lista[i]  
        heapify(lista, i, 0)  

lista = [12, 11, 13, 5, 6, 7]
print("Lista inițială:", lista)
heap_sort(lista)
print("Lista sortată:", lista)
